File: java/src/main/java/vsdk/toolkit/common/linealAlgebra/CoordinateSystem.java
package vsdk.toolkit.common.linealAlgebra;

public class CoordinateSystem {
    private Vector3D X = new Vector3D();
    private Vector3D Y = new Vector3D();
    private Vector3D Z = new Vector3D();

    public Vector3D getX() {
        return X;
    }

    public Vector3D getY() {
        return Y;
    }

    public Vector3D getZ() {
        return Z;
    }

    public void setX(Vector3D inX) {
        X = inX;
    }

    public void setY(Vector3D inY) {
        Y = inY;
    }

    public void setZ(Vector3D inZ) {
        Z = inZ;
    }

    public void setFromZAxis(Vector3D inZ) {
        Z = new Vector3D(inZ.x, inZ.y, inZ.z);

        float zz = (float)Math.sqrt(1.0f - inZ.z * inZ.z);
        if (zz < Numeric.EPSILON) {
            X.x = 1.0f;
            X.y = 0.0f;
            X.z = 0.0f;
        }
        else {
            X.x = inZ.y / zz;
            X.y = -inZ.x / zz;
            X.z = 0.0f;
        }

        Y.crossProduct(Z, X);
    }

    public void rectangularToSphericalCoord(Vector3D cIn, double[] phi, double[] theta) {
        if (phi == null || phi.length == 0 || theta == null || theta.length == 0) {
            throw new IllegalArgumentException("phi/theta output arrays must have length >= 1");
        }

        double z = cIn.dotProduct(Z);
        if (z > 1.0) {
            z = 1.0;
        }
        if (z < -1.0) {
            z = -1.0;
        }

        theta[0] = Math.acos(z);

        Vector3D c = new Vector3D();
        c.sumScaled(cIn, -z, Z);
        c.normalize(Numeric.EPSILON_FLOAT);

        double x = c.dotProduct(X);
        double y = c.dotProduct(Y);

        if (x > 1.0) {
            x = 1.0;
        }
        if (x < -1.0) {
            x = -1.0;
        }

        phi[0] = Math.acos(x);
        if (y < 0.0) {
            phi[0] = 2.0 * Math.PI - phi[0];
        }
    }

    public Vector3D sampleHemisphereCosTheta(double xi1, double xi2, double[] probabilityDensityFunction) {
        float phi = 2.0f * (float)Math.PI * (float)xi1;
        float cosPhi = (float)Math.cos(phi);
        float sinPhi = (float)Math.sin(phi);
        float cosTheta = (float)Math.sqrt(1.0 - xi2);
        float sinTheta = (float)Math.sqrt(xi2);

        Vector3D dir = new Vector3D();
        dir.combine(cosPhi, X, sinPhi, Y);
        dir.combine(sinTheta, dir, cosTheta, Z);

        if (probabilityDensityFunction != null && probabilityDensityFunction.length > 0) {
            probabilityDensityFunction[0] = cosTheta / Math.PI;
        }

        return dir;
    }

    public Vector3D sampleHemisphereCosNTheta(double n, double xi1, double xi2, double[] probabilityDensityFunction) {
        float phi = 2.0f * (float)Math.PI * (float)xi1;
        float cosPhi = (float)Math.cos(phi);
        float sinPhi = (float)Math.sin(phi);
        float cosTheta = (float)Math.pow(xi2, 1.0 / (n + 1.0));
        float sinTheta = (float)Math.sqrt(1.0 - cosTheta * cosTheta);

        Vector3D dir = new Vector3D();
        dir.combine(cosPhi, X, sinPhi, Y);
        dir.combine(sinTheta, dir, cosTheta, Z);

        if (probabilityDensityFunction != null && probabilityDensityFunction.length > 0) {
            probabilityDensityFunction[0] = (n + 1.0) * Math.pow(cosTheta, n) / (2.0 * Math.PI);
        }

        return dir;
    }
}

File: typescript/src/main/vsdk/toolkit/common/linealAlgebra/CoordinateSystem.ts
import { Numeric } from "./Numeric";
import { Vector3D } from "./Vector3D";

export class CoordinateSystem {
  private X: Vector3D = new Vector3D();
  private Y: Vector3D = new Vector3D();
  private Z: Vector3D = new Vector3D();

  public getX(): Vector3D {
    return this.X;
  }

  public getY(): Vector3D {
    return this.Y;
  }

  public getZ(): Vector3D {
    return this.Z;
  }

  public setX(inX: Vector3D): void {
    this.X.copy(inX);
  }

  public setY(inY: Vector3D): void {
    this.Y.copy(inY);
  }

  public setZ(inZ: Vector3D): void {
    this.Z.copy(inZ);
  }

  public setFromZAxis(inZ: Vector3D): void {
    this.Z.copy(inZ);

    const zz = globalThis.Math.sqrt(1.0 - inZ.z * inZ.z);
    if (zz < Numeric.EPSILON) {
      this.X.x = 1.0;
      this.X.y = 0.0;
      this.X.z = 0.0;
    }
    else {
      this.X.x = inZ.y / zz;
      this.X.y = -inZ.x / zz;
      this.X.z = 0.0;
    }

    this.Y.crossProduct(this.Z, this.X);
  }

  public rectangularToSphericalCoord(cIn: Vector3D, phi: number[], theta: number[]): void {
    if (phi === null || phi.length === 0 || theta === null || theta.length === 0) {
      throw new Error("phi/theta output arrays must have length >= 1");
    }

    let z = cIn.dotProduct(this.Z);
    if (z > 1.0) {
      z = 1.0;
    }
    if (z < -1.0) {
      z = -1.0;
    }

    theta[0] = globalThis.Math.acos(z);

    const c = new Vector3D();
    c.sumScaled(cIn, -z, this.Z);
    c.normalize(Numeric.EPSILON_FLOAT);

    let x = c.dotProduct(this.X);
    const y = c.dotProduct(this.Y);

    if (x > 1.0) {
      x = 1.0;
    }
    if (x < -1.0) {
      x = -1.0;
    }

    phi[0] = globalThis.Math.acos(x);
    if (y < 0.0) {
      phi[0] = 2.0 * globalThis.Math.PI - phi[0];
    }
  }

  public sampleHemisphereCosTheta(xi1: number, xi2: number, probabilityDensityFunction: number[] | null): Vector3D {
    const phi = 2.0 * globalThis.Math.PI * xi1;
    const cosPhi = globalThis.Math.cos(phi);
    const sinPhi = globalThis.Math.sin(phi);
    const cosTheta = globalThis.Math.sqrt(1.0 - xi2);
    const sinTheta = globalThis.Math.sqrt(xi2);

    const dir = new Vector3D();
    dir.combine(cosPhi, this.X, sinPhi, this.Y);
    dir.combine(sinTheta, dir, cosTheta, this.Z);

    if (probabilityDensityFunction !== null && probabilityDensityFunction.length > 0) {
      probabilityDensityFunction[0] = cosTheta / globalThis.Math.PI;
    }

    return dir;
  }

  public sampleHemisphereCosNTheta(n: number, xi1: number, xi2: number, probabilityDensityFunction: number[] | null): Vector3D {
    const phi = 2.0 * globalThis.Math.PI * xi1;
    const cosPhi = globalThis.Math.cos(phi);
    const sinPhi = globalThis.Math.sin(phi);
    const cosTheta = globalThis.Math.pow(xi2, 1.0 / (n + 1.0));
    const sinTheta = globalThis.Math.sqrt(1.0 - cosTheta * cosTheta);

    const dir = new Vector3D();
    dir.combine(cosPhi, this.X, sinPhi, this.Y);
    dir.combine(sinTheta, dir, cosTheta, this.Z);

    if (probabilityDensityFunction !== null && probabilityDensityFunction.length > 0) {
      probabilityDensityFunction[0] = (n + 1.0) * globalThis.Math.pow(cosTheta, n) / (2.0 * globalThis.Math.PI);
    }

    return dir;
  }
}
